common, all: polish a lot of hex handling code

karalabe opened the pull request:
This PR polishes our internal APIs a bit, notably makes the `common.IsHex` check a lot stricter by actually checking individual characters too, not just a general form (length, prefix) of a string for hex validation.

This is needed because we have a few other methods (e.g. `common.IsHexAddress`) that relied on `IsHex` to seemingly validate the string contents, but in reality the validation never happened. This isn't a problem in our current code base, but it could be a failure ready to happen, especially if you build on top of the go-ethereum code.

I've also added a ton of tests that all fail on the previous version, accepting invalid hex strings.

robotally commented:
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Tue Mar 1 10:04:49 UTC 2016_

fjl commented on common/bytes_test.go:
```diff
@@ -83,16 +83,45 @@ func (s *BytesSuite) TestCopyBytes(c *checker.C) {
+		{"", false},
+		{"0", false},
+		{"00", false},
+		{"0x", false},
```
Why is `0x` invalid?

karalabe replied:
That was the logic of the code previously. I didn't want to relax any checks performed.

fjl commented:
:+1:

obscuren commented on common/bytes.go:
```diff
@@ -122,9 +122,24 @@ func HasHexPrefix(str string) bool {
+		return false
+	}
+	// Verify each of the digits
+	for _, digit := range str[2:] {
```
Why not use regex?

obscuren replied:
`/(0[xX])?[A-Fa-f0-9]+/`

obscuren commented:
:-1: I don't quite understand why the `IsHex` is using some custom hex checking. We can replace this with regex quite easily which reduces the current ~20 LOC to 2 LOC.

obscuren commented:
+ I understand that the previous version of `IsHex` wasn't using RegEx either. Now that you want to patch it up proper, please do so.

Fredjet commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  fredjet15@gmail.com
    Domain ventureacceleration.com has exceeded the max emails per hour (50/50 (100%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext3.iad.github.net ([192.30.252.194]:51356 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by eaglepeak.ovasp.com with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.86)
	(envelope-from <noreply@github.com>)
	id 1aark3-00087Q-Dg
	for fred@ventureacceleration.com; Tue, 01 Mar 2016 16:22:04 -0500
Date: Tue, 01 Mar 2016 13:21:21 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1456867282;
	bh=LMs88kD0SHQmFzMvdlp7z9IDqPdsmY89Ct9v7WoBsO0=;
	h=From:Reply-To:To:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=ZZI+37XZJzwkhSdsUAUAqFpAAxLK/IifSXTaPIctZDyssMi0xcbrNrm7OjL1hgMht
	 xn8+KKaT4kle4QTQ3hsHq/dDuqtCN5X9upqGR1A1u5F+ofS6qEvHDXCSRz7jqVfVcO
	 OuWt1TURd0KrIK/aysFKJ4yKp83YgCPFzNJpPSHE=
From: Jeffrey Wilcke <notifications@github.com>
Reply-To: ethereum/go-ethereum <reply@reply.github.com>
To: ethereum/go-ethereum <go-ethereum@noreply.github.com>
Message-ID: <ethereum/go-ethereum/pull/2275/c190910735@github.com>
In-Reply-To: <ethereum/go-ethereum/pull/2275@github.com>
References: <ethereum/go-ethereum/pull/2275@github.com>
Subject: Re: [go-ethereum] common: make internal IsHex check stricter (digits)
 (#2275)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_56d607d1f1493_40353fdd4d6072b885975a";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: obscuren
X-GitHub-Recipient: Fredjet
List-ID: ethereum/go-ethereum <go-ethereum.ethereum.github.com>
List-Archive: https://github.com/ethereum/go-ethereum
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+0007d41275cfffb74330aaeb650a2db3f527dc73fb4c964492cf0000000112edc9d192a169ce08328280@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AAfUEvh8KK4Zt3Yl-tH5qeuN8uB8co4yks5ppK3RgaJpZM4HmT9K>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: fred@ventureacceleration.com


----==_mimepart_56d607d1f1493_40353fdd4d6072b885975a
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

+ I understand that the previous version of `IsHex` wasn't using RegEx either. Now that you want to patch it up proper, please do so.

---
Reply to this email directly or view it on GitHub:
https://github.com/ethereum/go-ethereum/pull/2275#issuecomment-190910735
----==_mimepart_56d607d1f1493_40353fdd4d6072b885975a
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<ul>
<li>I understand that the previous version of <code>IsHex</code> wasn't using RegEx either. Now that you want to patch it up proper, please do so.</li>
</ul>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br>Reply to this email directly or <a href="https://github.com/ethereum/go-ethereum/pull/2275#issuecomment-190910735">view it on GitHub</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AAfUEo2VaorwLlePcDmnEAnn2CT5R506ks5ppK3RgaJpZM4HmT9K.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/ethereum/go-ethereum/pull/2275#issuecomment-190910735"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

----==_mimepart_56d607d1f1493_40353fdd4d6072b885975a--

obscuren commented:
Additionally this will fail on `0X00` which is still valid hex.

Fredjet commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  fredjet15@gmail.com
    Domain ventureacceleration.com has exceeded the max emails per hour (55/50 (110%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext7.iad.github.net ([192.30.252.198]:41212 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by eaglepeak.ovasp.com with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.86)
	(envelope-from <noreply@github.com>)
	id 1aarlV-0008AR-TA
	for fred@ventureacceleration.com; Tue, 01 Mar 2016 16:23:34 -0500
Date: Tue, 01 Mar 2016 13:22:53 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1456867373;
	bh=d+666Vv6z9r76j/rsLOt/ZluxpTxeRuL9Okaq5lx/ss=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=2N2XKAT7/I9WQpSHkHSma7twDkoXlAANBYLAO2L4NZDfDCiSlOvbZMvngT15uBWW8
	 QIfNowf4v0m9mcY/hXnW/TSaiXUQ3xjcndSNLWfnCrEhTIVm0kZ/IovrThKXc0e9WA
	 goBt8AUd0e2b5Qnt2b1AoC7w9fTM0OOHrQzwNl7c=
From: Jeffrey Wilcke <notifications@github.com>
Reply-To: ethereum/go-ethereum <reply@reply.github.com>
To: ethereum/go-ethereum <go-ethereum@noreply.github.com>
Cc: Fred Sullivan <fred@ventureacceleration.com>
Message-ID: <ethereum/go-ethereum/pull/2275/c190911176@github.com>
In-Reply-To: <ethereum/go-ethereum/pull/2275@github.com>
References: <ethereum/go-ethereum/pull/2275@github.com>
Subject: Re: [go-ethereum] common: make internal IsHex check stricter (digits)
 (#2275)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_56d6082d62f6d_40623fdd4d6072b84420a3";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: obscuren
X-GitHub-Recipient: Fredjet
X-GitHub-Reason: mention
List-ID: ethereum/go-ethereum <go-ethereum.ethereum.github.com>
List-Archive: https://github.com/ethereum/go-ethereum
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+0007d4122fce91e7a2e3ff417eae19500cc3cccd5665419f92cf0000000112edca2d92a169ce08328280@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AAfUEgu5i3VHTuStN2mZtTgLOUuNnaK_ks5ppK4tgaJpZM4HmT9K>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: fred@ventureacceleration.com


----==_mimepart_56d6082d62f6d_40623fdd4d6072b84420a3
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

Additionally this will fail on `0X00` which is still valid hex. 

---
Reply to this email directly or view it on GitHub:
https://github.com/ethereum/go-ethereum/pull/2275#issuecomment-190911176
----==_mimepart_56d6082d62f6d_40623fdd4d6072b84420a3
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p>Additionally this will fail on <code>0X00</code> which is still valid hex. </p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br>Reply to this email directly or <a href="https://github.com/ethereum/go-ethereum/pull/2275#issuecomment-190911176">view it on GitHub</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AAfUEiDesItexaxF4kFDVUiWhTTdIrPjks5ppK4tgaJpZM4HmT9K.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/ethereum/go-ethereum/pull/2275#issuecomment-190911176"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

----==_mimepart_56d6082d62f6d_40623fdd4d6072b84420a3--

karalabe commented:
I wanted to avoid regexes since they are quite heavy and probably would take significantly more time to verify an address than it takes now. Then again, a hex check will still seem instantaneous and insignificant compared to any other operation that we do, so it should really be fine to use it in this scenario. Will update in a moment.

karalabe commented:
@obscuren Do we want to support `0X` too? Because that would require quite a lot of rewrite as all the parsers use strictly `0x` only.

karalabe commented:
@fjl @obscuren I've pushed a new version based on regular expressions. The regexp used was `0x([0-9a-fA-F]{2})+`:

 * Mandatory lowercase `0x` prefix
 * Even number of hex characters are required
 * The empty hex string is not valid (i.e. use `0x00`, not `0x`)

The reason for the above pattern is to retain the current semantics of the method for this PR. I concur that we could change the `common` methods to support capital `0X`, possibly support the empty `0x` and maybe/maybe not an optional prefix, but those could lead to unforeseen consequences so I'd like to postpone that to post 1.4 so that we notice if something goes belly up.

obscuren commented:
We should support the optional prefix and `0X` since we're fixing it now anyway.

karalabe commented:
K, will put it back into progress and start doing a more involved fix.

karalabe commented:
@obscuren @fjl I've made some significant changes per requested, please re-review.

Notably:

 * Updated `common.IsHex` to verify the hex content too, not just the prefix.
 * Updated `common` to accept both `0x` as well as `0X` prefixes.
 * Deleted `common.Bytes2Hex` as it was an alias to `hex.EncodeToString`.
 * Split `common.ToHex` into `common.NumberToHex` (which needs to encode the empty byte slice to `0x0`) and `common.BytesToHex` (which needs to encode the empty byte slice to `0x`). We did the latter wrong until now, potentially causing an `FromHex(ToHex)` to produce bad results.
 * Added an error log into `common.Hex2Bytes` and `commonHex2BytesFixed` if the input data was a malformed string (which we currently silently ignored and returned nil as the "parsed" data).

The remained of the PR are the modifications to the rest of the code needed to update them to the modified API.

fjl commented on common/bytes.go:
```diff
@@ -225,22 +234,14 @@ func LeftPadString(str string, l int) string {
-
-	return zeros + str
-
+	return hex.EncodeToString(make([]byte, (l-len(str))/2)) + str
```
Consider using `strings.Repeat`.

fjl commented on common/bytes.go:
```diff
@@ -23,29 +23,38 @@ import (
+}
+
+// NumberToHex converts a byte slice to hexadecimal notation, prefixed with 0x,
+// with the added rule that the empty slice is encoded as the zero value.
```
"encoded as the zero value" -> "encoded as "0x0"

fjl commented on common/bytes.go:
```diff
@@ -117,28 +126,29 @@ func CopyBytes(b []byte) (copiedBytes []byte) {
+	h, err := hex.DecodeString(str)
+	if err != nil {
+		glog.V(logger.Error).Infof("Invalid hex string to decode: %s: %v", str, err)
+	}
```
I don't agree with error logging in a primitive conversion function.
If the caller wants to know whether the input is valid, it can use hex.DecodeString
and handle the error.

fjl commented on common/bytes.go:
```diff
@@ -23,29 +23,38 @@ import (
+// BytesToHex converts a byte slice to hexadecimal notation, prefixed with 0x.
+func BytesToHex(blob []byte) string {
+	return "0x" + hex.EncodeToString(blob)
+}
```
Can we name this `Bytes0x`? I think it needs to be obvious in the name what the difference to hex.EncodeToString is.

fjl commented:
@karalabe do you still want this merged?

karalabe commented:
Yeah, I'll get the console in first then redo this one.

fjl commented:
I'm closing this now because it's very old and we need to discuss first. I have a similar change that moves all hex processing to a separate package. I'll submit it along with some other cleanup changes to package common.
